interfaces/builtin/mount_control.go

```diff
@@ -0,0 +1,397 @@
+}
+
+// mountControlInterface allows creating transient and persistent mounts
+type mountControlInterface struct {
+	commonInterface
+}
+
+// We prohibit expressions containing a comma, as that is used by AppArmor to
+// separate rules; otherwise, one could specify a path like
+//   $SNAP_DATA/foo, /** rw
+// and get read-write access to the whole filesystem.
+var whereRegexp = regexp.MustCompile(`^(/media|\$SNAP_COMMON|\$SNAP_DATA)/[^,]*$`)
+
+// Excluding spaces and other characters which might allow constructing a
+// malicious string like
+//   auto) options=() /malicious/content /var/lib/snapd/hostfs/...,\n mount fstype=(
+var typeRegexp = regexp.MustCompile(`^[a-z0-9-]+$`)
+
+type MountInfo struct {
+	what       string
+	where      string
+	persistent bool
+	types      []string
+	options    []string
+}
+
+func parseStringList(mountEntry map[string]interface{}, fieldName string) ([]string, error) {
+	var list []string
+	value, ok := mountEntry[fieldName]
+	if ok {
+		interfaceList, ok := value.([]interface{})
+		if !ok {
+			return nil, fmt.Errorf(`mount-control "%s" must be an array of strings (got %q)`, fieldName, value)
+		}
+		for i, iface := range interfaceList {
+			valueString, ok := iface.(string)
+			if !ok {
+				return nil, fmt.Errorf(`mount-control "%s" element %d not a string (%q)`, fieldName, i+1, iface)
+			}
+			list = append(list, valueString)
+		}
+	}
+	return list, nil
+}
+
+func enumerateMounts(plug interfaces.Attrer, fn func(mountInfo *MountInfo) error) error {
+	mountAttr, ok := plug.Lookup("mount")
+	if !ok {
+		return nil
+	}
+	mounts, ok := mountAttr.([]interface{})
+	if !ok {
+		return mountAttrTypeError
+	}
+
+	for _, m := range mounts {
+		mount, ok := m.(map[string]interface{})
+		if !ok {
+			return mountAttrTypeError
+		}
+
+		what, ok := mount["what"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "what" must be a string`)
+		}
+
+		where, ok := mount["where"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "where" must be a string`)
+		}
+
+		persistent := false
+		persistentValue, ok := mount["persistent"]
+		if ok {
+			if persistent, ok = persistentValue.(bool); !ok {
+				return fmt.Errorf(`mount-control "persistent" must be a boolean`)
+			}
+		}
+
+		types, err := parseStringList(mount, "type")
+		if err != nil {
+			return err
+		}
+
+		options, err := parseStringList(mount, "options")
+		if err != nil {
+			return err
+		}
+
+		mountInfo := &MountInfo{
+			what:       what,
+			where:      where,
+			persistent: persistent,
+			types:      types,
+			options:    options,
+		}
+
+		if err := fn(mountInfo); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validateWhatAttr(what string) error {
+	if what[0] != '/' {
+		return fmt.Errorf(`mount-control "what" attribute must be an absolute path`)
+	}
+
+	if !cleanSubPath(what) {
+		return fmt.Errorf(`mount-control "what" pattern is not clean: %q`, what)
+	}
+
+	if _, err := utils.NewPathPattern(what); err != nil {
```

Comment:
So to be clear, the `what` and `where` properties here are expected not to be traditional regular expressions but rather to be very explicitly AppArmor Regular Expressions (AARE)? 

I fear that this logic in this validation right here is not quite correct for a few reasons, because when we are validating the AARE here, we have not yet expanded the $SNAP, etc. variables. I think that in order for the validation to be meaningful, it should match the usage in AppArmorConnectedPlug and expand the variables first, then check that the resulting string is a valid AARE, but this brings up the issue that the logic inside ExpandSnapVariables doesn't seem to be intelligent enough to handle being in an AARE AFAICT, which leads me to the utterly terrible recommendation that we actually implement our own unified parsing function which replaces these environment variables concurrently while validating that it is a valid AARE. For example, look at what we do for the input `where: "$SNAP/\$SNAP/foo"`. the `\$SNAP` presumably should not be expanded at all to mean literally the (silly) file `/snap/hello-world/1/$SNAP/foo`, but the validation code here essentially ends up building the regexp `^\$SNAP/\$SNAP/foo$` but the expansion code means that we write out to the AppArmor profile the string `/snap/hello-world/45/\/snap/hello-world/45/foo`. Also more problematically is the fact that we don't actually quote the input strings for what/where at all, but we still allow various dangerous characters in the input string such as `,`, and the effect being that we can inject rules with via a `where: "$SNAP/foo, /** rw` sort of string. Granted we are generally protected against such attacks by this interface being super-privileged but still I'm just uncomfortable with the how much we are allowing without any sort of transformation. 

Given the above issues, I do legitimately wonder if the whole AARE validation is overkill, I wonder if we couldn't instead get away with _just_ supporting a single "*" glob character in the input string to mean just any sequence of characters except "/" (or some other very easy to understand and implement set of characters) so that you can write paths like this

```yaml
plugs:
  mount-control:
    - what: "/dev/sda*"
      where: $SNAP_COMMON/foo
```

And if folks need more complicated regular expressions they can just add onto the dictionary like in

```yaml
plugs:
  mount-control:
    - what: "/dev/sda*"
      where: $SNAP_COMMON/foo
    - what: "/dev/vda*"
      where: $SNAP_COMMON/foo
```

instead of something complicated like

```yaml
plugs:
  mount-control:
    - what: "/dev/{s,v}da*"
      where: $SNAP_COMMON/foo
```

And then we disallow dangerous characters like `,` or `(`, `{`, etc. (and we revisit the situation when someone comes along with a valid use case for including these characters in the what/where specifications)

Basically, I'm not sure if the cost/benefit is worth it to maintain AppArmor Regular Expression validation code here in snapd. in exchange for the ability for our users to write very specific mount restrictions. I'm also somewhat concerned about potential confusion about the syntax here, since folks will not be writing normal "full" regular expressions here, they are writing AppArmor Regular Expressions instead. I can certainly attest to having fallen victim to writing AARE incorrectly numerous times because I forgot that I am writing AARE and not normal RE. 

Clearly the disadvantage of my approach is that we lose the ability to be very specific in the `what` and `where` settings for the mount-control plug attributes, which maybe is the whole reason we implemented all this complexity is because we feel that we need the ability to be very specific in mount-control plug attribute specifications, but I would argue that the super-privilegeness of this interface means we can help users craft logical and "good enough" strings for this.

If we do want to move forward with allowing arbitrary and complex regular expression matching for what/where, I guess the direction I would approach it from is taking as input from what/where a traditional regular expression string and transform that into a quoted and safe AARE, failing on regular expression constructs which are not supported in AARE, but that seems like an incredible amount of work for IMHO very little reward.

cc @pedronis for opinions

Reply:
You are right that allowing full AARE on the plug specification is potentially risky, but we are already taking some precautions: the `where` is currently pre-validated like this:
```
// We prohibit expressions containing a comma, as that is used by AppArmor to
// separate rules; otherwise, one could specify a path like
//   $SNAP_DATA/foo, /** rw
// and get read-write access to the whole filesystem.
var whereRegexp = regexp.MustCompile(`^(/media|\$SNAP_COMMON|\$SNAP_DATA)/[^,]*$`)
```
which already has the unfortunate side-effect of disallowing the `{a,b}` patterns, since it prohibits using a comma (but it's not such a big deal, since once can specify `**` at any point). It's true, though, that we don't have any validation for the `what`.

But while we should indeed find a solution to this, I think that the fact that the interface is super-privileged means that we should not be blocked by this. It seems to me that the content interface is also taking the source and target paths provided by the developer and dumping them into apparmor rules, so I would say that all these (valid) issues you are pointing out here do already exist in a non privileged interface.

Reply:
OK, I added a [commit](https://github.com/snapcore/snapd/pull/10739/commits/80d50c4a7c579dfe1ce334c727060656bec41c27) which should  prevent using a comma inside a path expression; I would then propose that we make the regexp more permissive again, and re-allow the comma in there.

Reply:
And one more commit, not necessarily needed since the previous one aready protects against this, but one layer more won't harm: https://github.com/snapcore/snapd/pull/10739/commits/3da4c7300d8e6ed29fa58e13a50ea15effe9eeee